state/modelgeneration.go

```diff
@@ -172,6 +172,55 @@ func assignGenerationUnitTxnOps(id, appName, unitName string) []txn.Op {
 	}
 }
 
+// Complete marks the generation as completed.
+func (g *Generation) Complete() error {
+	buildTxn := func(attempt int) ([]txn.Op, error) {
+		if attempt > 0 {
+			if err := g.Refresh(); err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
+		if !g.Active() {
+			return nil, errors.New("next generation is not currently active")
+		}
+		ok, err := g.CanAutoComplete()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if !ok {
+			return nil, errors.New("next generation can not auto complete")
+		}
+		assignedUnits := g.AssignedUnits()
+		unitCnt := len(assignedUnits)
+		assertAndCriteria := []bson.D{
+			{{"active", true}},
+			{{"completed", 0}},
+			{{"assigned-units", bson.D{{"$size", unitCnt}}}},
+		}
+		// Ensure the criteria to determine auto complete don't change under us.
+		for app, units := range assignedUnits {
+			appField := fmt.Sprintf("assigned-units.%s", app)
+			assertAndCriteria = append(assertAndCriteria, bson.D{{appField, bson.D{{"$all", units}}}})
+		}
+		time, err := g.st.ControllerTimestamp()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		ops := []txn.Op{
+			{
+				C:      generationsC,
+				Id:     g.doc.Id,
+				Assert: bson.D{{"$and", assertAndCriteria}},
+				Update: bson.D{
+					{"$set", bson.D{{"completed", time.Unix()}}},
```

Comment:
@manadart, Is time what you had in mind for completed, once it's completed?  Or something else?

Reply:
Yes, exactly this.